stackhpc_ipa_hardware_managers/system_nic.py

```diff
@@ -0,0 +1,241 @@
+    relative_path = os.path.relpath(uevent_file, '/sys/class/net/')
+    return _get_base_in_relative_path(relative_path)
+
+
+def _get_ethtool_output(dev, **kwargs):
+    kwargs["shell"] = True
+    ethtool_output = ''
+    try:
+        ethtool_output, _e = utils.execute(
+            "ethtool -i {device}".format(device=dev),
+            **kwargs)
+    except (processutils.ProcessExecutionError, OSError):
+        raise errors.CleaningError("Unable to run ethtool")
+    return ethtool_output
+
+
+def _get_ethtool_field(ethtool_output, field):
+    for line in ethtool_output.splitlines():
+        (candidate_field, value) = tuple(line.split(": "))
+        if candidate_field == field:
+            return value
+
+
+def _get_expected_property(node, node_property):
+    try:
+        expected_property = node[node_property]
+    except KeyError:
+        raise errors.CleaningError(
+            "Expected property '{0}' not found. You should make sure all items"
+            " in the nic_firmware list contain {0}".format(node_property))
+    return expected_property
+
+
+def _get_nic_firmware_versions(vendor_id, device_id):
+    # there might be multiple identical cards, we must check them all
+    devices = _parse_lspci_output(_get_lspci_output(vendor_id, device_id))
+    interfaces = []
+    for device in devices:
+        interface_name = _pci_addr_to_net_interface(device['Slot'])
+        if interface_name is None:
+            raise errors.CleaningError(
+                "Could not determine network interface name. The pci_id was: "
+                "{vendor_id}:{device_id}. Does this correspond to a network "
+                "card?".format(vendor_id=vendor_id, device_id=device_id))
+        interfaces.append(interface_name)
+    return dict(map(lambda x: (
+        x, _get_ethtool_field(_get_ethtool_output(x), "firmware-version")),
+        interfaces))
+
+
+class SystemNICHardwareManager(hardware.HardwareManager):
+    """Checks firmware version for a given network card"""
+
+    HARDWARE_MANAGER_VERSION = '1.0'
+
+    def evaluate_hardware_support(self):
+        """Declare whether the system is supported by this manager.
+
+        :returns: HardwareSupport level for this manager.
+        """
+        # This should work for anything which supports ethtool, lspci
+        return hardware.HardwareSupport.SERVICE_PROVIDER
+
+    def get_clean_steps(self, node, ports):
+        """Get a list of clean steps with priority.
+
+        :param node: The node object as provided by Ironic.
+        :param ports: Port objects as provided by Ironic.
+        :returns: A list of cleaning steps, as a list of dicts.
+        """
+        return [{'step': 'verify_nic_firmware',
+                 'priority': 90,
+                 'interface': 'deploy',
+                 'reboot_requested': False,
+                 'abortable': True}]
+
+    def _get_lspci_output(*args, **kwargs):
+        return _get_lspci_output(*args, **kwargs)
```

Comment:
It might be more canonical to leave these here as static methods. I don't think the testing should look too different. Example:

https://github.com/stackhpc/stackhpc-monasca-agent-plugins/blob/master/stackhpc_monasca_agent_plugins/tests/unit/checks/test_slurm.py

Reply:
Yeah, you are right, that is mess, but I don't like the idea of mocking those methods as the test becomes dependent on the implementation. What I propose is to test the parsing separately and make the SystemNICHardwareManager tests completely independent of the parsing logic (assume it works). Thoughts?

Reply:
Yeah, fair point about not making the tests depend on the implementation / testing private methods.

Reply:
nit: I'm fine with using global functions, but why not call them directly? Looks like its to enable testing using a mock. It's pretty common practice in python (and OpenStack in particular) to use the mock library to do that sort of thing, and it allows you to check error paths. The test is dependent upon the implementation anyway due to the mock object.